Another auto rollover

da-kami opened the pull request:
After discussion with @thomaseizinger we decided to put the rollover logic on the cfd and unit-test it. 

Auto-rollover actually only depends on the cfd expiry. We hardcode the roll-over offset as `1 hour` for the moment. The unit-tests should cover all bounds and should help understand the logic. (If not please scream :)
Auto-rollover itself is not tested as part of the actor tests. The unit tests should cover the logic. The rest is just wiring things together and should not require testing.

We are adding an actor test for `ProposeRollover` (to test the logic related to rollover itself - accept, reject...) in a follow up PR. This test will be triggered through sending the `ProposeRollover` message to the actor in the tests. 
In the production code we don't send this message to the cfd actor from the outside (it is triggered internally upon auto-rollover).

klochowicz committed:
Don't use tokio feature from xtra

Prefer using our FutureExt to enforce keeping RemoteHandles to prevent zombie
tasks.

klochowicz committed:
Tie lifetimes of spawned tasks to actors

klochowicz committed:
Disallow calls to tokio::spawn outside dedicated extension trait

Prevent accidentally spawning a task that could outlive the actor system
and become a zombie.

Any long-lived task should be tied to its owner by keeping RemoteHandle around.

klochowicz committed:
Expand Tasks API for ergonomics

The fact that we're storing a Vec<RemoteHandle<()>) internally is an
implementation detail.
Calling `spawn_with_handle` is done internally now, guarded by a `debug_assert!`
macro discouraging from doing it twice.

da-kami committed:
Cleanup impl blocks

da-kami committed:
Fixed settlement interval

A configurable settlement interval for the maker causes friction when implementing auto-rollover, because the settlement interval is not known to the taker.
The settlement interval should not be configurable, thus the configurable settlement interval is removed on the maker side.

The cfd actors and oracle actor use a global constant to decide the settlement interval and oracle even lookahead.
Not pulling the constant into the actors is a conscious decision, it allows us to set this to a different value in the actor tests.
The fact that we use the same constant for cfd actor and oracle actor is an implementation detail.

da-kami committed:
Remove rollover from `CfdActions`

Rollover is not triggerable from the outside anymore, but also not automated yet.

da-kami committed:
Auto-rollover cfds

Triggered every 5 minutes.
We decide to roll-over based on the `expiry_timestamp` of the cfd (which is the oracle attestation timestamp).

If we are within `23 hour until the expiry timestamp is reached` and we did not `receive an attestation` yet, we propose a rollover.

Note: `within 23 hour until the expiry timestamp is reached` should actually be enough as condition, but for completion we added `receive an attestation` as well.
If we received an attestation we are sure we don't want to roll-over anymore.

Co-authored-by: rishflab <rishflab@hotmail.com>

thomaseizinger committed:
Remove duplicate message

This is already logged in the connection actor.

da-kami committed:
Move tests to bottom of file

thomaseizinger committed:
Transition uses of `do_send` to `send`

`do_send` on `MessageChannel` suffers from the same problems as
`Address#do_send` and `Address#do_send_async`.

Our uses only send messages that never fail so we don't need to adjust
anything.

thomaseizinger committed:
Ban use of `do_send` function via clippy

da-kami committed:
fixup! Auto-rollover cfds

Send `ProposeRollover` message to cfd actor from `AutoRollover` to trigger the proposal through message rather than function call.

da-kami commented on daemon/src/taker_cfd.rs:
```diff
@@ -326,6 +329,51 @@ impl<O, M, W> Actor<O, M, W> {
+        let cfds = load_all_cfds(&mut conn).await?;
+
+        // cleanup all pending proposals because auto-rollover will create a new one
+        self.current_pending_proposals.clear();
```
Note for tomorrow Daniel: Is this actually needed given that we filter?

da-kami replied:
Keeping this statement here means that we clean up rollovers every 5 minutes. I think this is fine. We could also just do it in case we have a proposal. I don't think it should do any harm to do this here - what do you think @thomaseizinger ?

bonomat commented on daemon/src/model/cfd.rs:
```diff
@@ -1274,6 +1274,31 @@ impl Cfd {
+        let remaining_till_expiry = self.expiry_timestamp() - now;
+        let rollover_window = Duration::hours(1);
+
+        if matches!(
```
My personal preference is to do an early return instead of if/else, e.g. 

```
if !matches!(self.state.clone(),
            CfdState::Open {
                attestation: None,
                ..
            })
{
return None;
}
```

thomaseizinger replied:
Second this!

da-kami replied:
this changed completely, I think you will like the change :)

bonomat reviewed (approved):
Good work 👍

da-kami committed:
fixup! Auto-rollover cfds

Improve logging

thomaseizinger commented on daemon/src/lib.rs:
```diff
@@ -52,6 +52,8 @@ pub const HEARTBEAT_INTERVAL: std::time::Duration = Duration::from_secs(5);
 
 pub const N_PAYOUTS: usize = 200;
 
+pub const SETTLEMENT_INTERVAL: time::Duration = time::Duration::hours(24);
```
Some docs would be nice on this!

da-kami replied:
done! let me know if that's what you had in mind

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -760,15 +760,10 @@ mod tests {
 
     impl Order {
-        fn dummy() -> Self {
+        fn dummy_db() -> Self {
```
I don't understand the value of this rename?

da-kami replied:
The problem is, that the impl blocks in the tests are accessible over all tests. So, if we have a `dummy` function on `Order` / `Cfd` then that dummy will be accessible from any test.
To make it clearer which dummy is used where I started being more explicit about the dummies. Keep the `dummy_...` function private (because that is specific to a test block and should not be re-used across tests because it could cause weird side effects if we change the dummy...) and be more explicit about the name.

We don't have to do this rename, but I would make it a convention to add where the dummy is used.

thomaseizinger commented on daemon/src/lib.rs:
```diff
@@ -298,6 +303,10 @@ where
         tasks.add(oracle_ctx.run(oracle_constructor(cfds, Box::new(fan_out_actor))));
 
+        cfd_actor_addr
+            .do_send_async(taker_cfd::AutoRollover)
```
Why is this usage not flagged by clippy :cry: 

Are you rebased onto latest master?

thomaseizinger replied:
Needs to use `send`.

da-kami replied:
When I wrote this code we still had `do_send_async` `Sync` messages in `lib.rs` - I thought this was a concious decision to not block the startup. 
We did rebase on `master` yesterday, but the removal of `do_send_async` + clippy rule landed later. I'll fix it :)

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -351,7 +353,7 @@ impl<O, M, W> Actor<O, M, W> {
 impl<O: 'static, M: 'static, W: 'static> Actor<O, M, W> {
     async fn handle_auto_rollover(&mut self, ctx: &mut Context<Self>) -> Result<()> {
-        tracing::debug!("Auto rollover");
+        tracing::trace!("Auto rollover message received");
```
I _think_ we have a general `trace` log already for message that we receive.

da-kami replied:
I'll rebase on latest master and see. I just ran it and did not see a log for these messages besides the logs I added.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -1666,179 +1869,196 @@ mod tests {
-    #[serde(with = "::bdk::bitcoin::util::amount::serde::as_sat")]
-    pub taker_lock_amount: Amount,
+        assert!(cfd.rollover_proposal().is_some());
+    }
```
I think this test should be `none`.

da-kami committed:
Remove daemon `clippy.toml`

da-kami committed:
fixup! Auto-rollover cfds

Do not use `do_send_async` for `AutoRollover` sync.

da-kami committed:
fixup! Auto-rollover cfds

Use `do_send_async` to trigger sending message to the actor itself to avoid deadlock.

thomaseizinger committed:
fixup! Transition uses of `do_send` to `send`

da-kami committed:
fixup! Auto-rollover cfds

Fix logic and more extensive tests

da-kami committed:
fixup!  Fixed settlement interval

Comment to explain constant.

da-kami committed:
fixup! Auto-rollover cfds

Move trigger for Autorollover after connection is established, otherwise actor is blocked
